plugins/rbac-backend/src/service/policies-rest-api.ts

```diff
@@ -241,54 +228,60 @@ export class PolicesServer {
 
         const entityRef = this.getEntityReference(request);
 
-        const oldPolicyRaw = request.body.oldPolicy;
-        if (!oldPolicyRaw) {
+        const oldPolicyRaw: RoleBasedPolicy[] = request.body.oldPolicy;
+        if (isEmpty(oldPolicyRaw)) {
           throw new InputError(`'oldPolicy' object must be present`); // 400
         }
-        const newPolicyRaw = request.body.newPolicy;
-        if (!newPolicyRaw) {
+        const newPolicyRaw: RoleBasedPolicy[] = request.body.newPolicy;
+        if (isEmpty(newPolicyRaw)) {
           throw new InputError(`'newPolicy' object must be present`); // 400
         }
 
-        oldPolicyRaw.entityReference = entityRef;
-        let err = validatePolicy(oldPolicyRaw);
-        if (err) {
-          throw new InputError( // 400
-            `Invalid old policy object. Cause: ${err.message}`,
-          );
-        }
-        newPolicyRaw.entityReference = entityRef;
-        err = validatePolicy(newPolicyRaw);
-        if (err) {
-          throw new InputError( // 400
-            `Invalid new policy object. Cause: ${err.message}`,
-          );
-        }
+        [...oldPolicyRaw, ...newPolicyRaw].forEach(element => {
+          element.entityReference = entityRef;
+        });
 
-        const oldPolicy = this.transformPolicyToArray(oldPolicyRaw);
-        const newPolicy = this.transformPolicyToArray(newPolicyRaw);
+        const processedOldPolicy = await this.processPolicies(
+          oldPolicyRaw,
+          true,
+          'old policy',
+        );
 
-        if (await this.enforcer.hasPolicy(...newPolicy)) {
-          if (isEqual(oldPolicy, newPolicy)) {
-            response.status(204).end();
-            return;
-          }
-          throw new ConflictError(); // 409
-        }
+        oldPolicyRaw.sort((a, b) =>
+          a.permission === b.permission
+            ? this.nameSort(a.policy!, b.policy!)
+            : this.nameSort(a.permission!, b.permission!),
+        );
 
-        if (!(await this.enforcer.hasPolicy(...oldPolicy))) {
-          throw new NotFoundError(); // 404
+        newPolicyRaw.sort((a, b) =>
+          a.permission === b.permission
+            ? this.nameSort(a.policy!, b.policy!)
+            : this.nameSort(a.permission!, b.permission!),
+        );
+
```

Comment:
I guess, we need to apply also check, that oldPolicy.length and newPolicy.length are equal.

Reply:
Wouldn't the `isEqual` check [below](https://github.com/janus-idp/backstage-plugins/pull/984/files#diff-827c86ea2bf7aac8f0f33a83e06805cde96c6b71eaadaebfcd1880cd97cc35ebR263) already account for verifying whether `oldPolicyRaw.length` and `newPolicyRaw.length` are equal?

Or are you requesting an additional check to return a more specific error?

Reply:
I have added an additional [check](https://github.com/janus-idp/backstage-plugins/pull/984/files#diff-827c86ea2bf7aac8f0f33a83e06805cde96c6b71eaadaebfcd1880cd97cc35ebR267-R271) to verify whether the oldPolicy has more permissions than the newPolicy. Throws an error if this event is true.